### PR TITLE
Aligning to the PR of Daniel Rosenwasser

### PR DESCRIPTION
**Pull request by lhk**
I tried to incorporate the changes @DanielRosenwasser made in his PR.
The defininitions now pass both tests.

You had a lot of structure in your definitions that wasn't present in my code. I copied your code for the animations.
There is still some functionality missing in my definitions file: For example the ParseCommands interface for path parsing. How did you create that ? It is not really documented and the source code of snap is hard to read. I couldn't say wether the parsed commands are numbers or strings.

That being said, I think the two versions are now comparable. It is probably a matter of taste which structure is preferable. Therefore it would be great if you could issue your PR again. You seem to have a deeper understanding of Snap than me (see ParsedCommands) and the users would probably profit from this.

### CODE REVIEW

**Comment by lhk** [comment]
I found a good snapsvg tutorial which I used to expand the testsuite:
http://svg.dabbles.info/

This tutorial covers very many aspects of Snap. I have contacted the author, if he would allow me to add his code as examples to this repo.

With this tutorial series, I could remove some errors from the definitions.

**Comment by lhk** [comment]
The author of the tutorial has answered my mail, I've committed a test file with snippets of his tutorial

**Comment by vvakame** [comment]
@lhk thank you for your contributions!
please resolve conflict.

**Comment by lhk** [comment]
resolved

File: snapsvg/snapsvg-tests-2.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
+/// <reference path="snapsvg.d.ts" />
+
+// tests by Daniel Rosenwasser
```
Comment (DanielRosenwasser) [inline snapsvg/snapsvg-tests-2.ts]: Could you make this

```TypeScript
// Tests by: Daniel Rosenwasser <https://github.com/DanielRosenwasser>
```

Thanks!

File: snapsvg/snapsvg.d.ts [hunk 2]
```diff
@@ -13,6 +13,28 @@ declare module mina {
         status: Function;
         stop: Function;
     }
+	export interface AnimationDescriptor{
+		id: string;
```
Comment (DanielRosenwasser) [inline snapsvg/snapsvg.d.ts]: The rest of the file uses spaces, this is two tabs

File: snapsvg/snapsvg.d.ts [hunk 3]
```diff
@@ -51,13 +73,14 @@ declare module Snap {
...
-	export function animate(from:Array<number>,to:number,setter:Function,duration:number,easing:Function,callback:Function):mina.Mina;
-	export function animate(from:number,to:Array<number>,setter:Function,duration:number,easing:Function,callback:Function):mina.Mina;
-	export function animate(from:Array<number>,to:Array<number>,setter:Function,duration:number,easing:Function,callback:Function):mina.Mina;
-	export function animation(attr:Object,duration:number,easing?:Function,callback?:Function):Object;
+	export function animate(from:number,to:number,setter:Function,duration:number,easing?:(num:number)=>number,callback?:()=>void):mina.MinaAnimation;
+	export function animate(from:Array<number>,to:number,setter:Function,duration:number,easing?:(num:number)=>number,callback?:()=>void):mina.MinaAnimation;
+	export function animate(from:number,to:Array<number>,setter:Function,duration:number,easing?:(num:number)=>number,callback?:()=>void):mina.MinaAnimation;
+	export function animate(from:Array<number>,to:Array<number>,setter:Function,duration:number,easing?:(num:number)=>number,callback?:()=>void):mina.MinaAnimation;
```
Comment (DanielRosenwasser) [inline snapsvg/snapsvg.d.ts]: I would use union types here (e.g. `from: number | number[]`)

Comment (DanielRosenwasser) [inline snapsvg/snapsvg.d.ts]: I'd also rename `setter` to `updater` and make its type `(n: number) => void`

File: snapsvg/snapsvg.d.ts [hunk 4]
```diff
@@ -111,33 +147,48 @@ declare module Snap {
...
+        string: string;
+        globalMatrix: Snap.Matrix;
+        localMatrix: Snap.Matrix;
+        diffMatrix: Snap.Matrix;
+        global: string;
+        local: string;
+        toString(): string;
+    }
+	export interface Animation{
+		attr:Object;
```
Comment (DanielRosenwasser) [inline snapsvg/snapsvg.d.ts]: I think this should be `{ [attr: string]: string | number | boolean }`

File: snapsvg/snapsvg.d.ts [hunk 5]
```diff
@@ -111,33 +147,48 @@ declare module Snap {
...
+		add(el:Snap.Element):Snap.Element;
 		addClass(value:string):Snap.Element;
 		after(el:Snap.Element):Snap.Element;
-		animate(attrs:Object,duration:number,easing?:Function,callback?:Function):Snap.Element;
+		animate(attrs:Object,duration:number,easing?:(num: number)=> number,callback?:()=>void):Snap.Element;
 		animate(animation:any):Snap.Element;
 		append(el:Snap.Element):Snap.Element;
 		appendTo(el:Snap.Element):Snap.Element;
-		asPX(attr:string,value?:string):Snap.Element;
+		asPX(attr:string,value?:string):any;            //TODO: check what is really returned
```
Comment (DanielRosenwasser) [inline snapsvg/snapsvg.d.ts]: It returns a `number`

File: snapsvg/snapsvg.d.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline snapsvg/snapsvg.d.ts]: Let's just normalize on 4 spaces per tab in this file

File: snapsvg/snapsvg.d.ts [hunk 6]
```diff
@@ -155,70 +206,95 @@ declare module Snap {
...
+	
 	interface Paper extends Snap.Element {
 
 	clear():void;
 	el(name:string, attr:Object):Snap.Element;
+	filter(filstr:string):Snap.Element;
 	gradient(gradient:string):any;
-	g(varargs?:any):any;
-	group(el?:any,...els:any[]):any;
+	g(varargs?:any):Object;
```
Comment (DanielRosenwasser) [inline snapsvg/snapsvg.d.ts]: This should have the same type as `group` from my understanding

File: snapsvg/snapsvg.d.ts [hunk 7]
```diff
@@ -155,70 +206,95 @@ declare module Snap {
...
+	group(...els:any[]):any;
```
Comment (DanielRosenwasser) [inline snapsvg/snapsvg.d.ts]: This seems to be an array of `Element`

File: snapsvg/snapsvg.d.ts [hunk 8]
```diff
@@ -234,13 +310,14 @@ declare module Snap {
...
-	text(x:number,y:number,text:string):Snap.Element;
-	text(x:number,y:number,text:Array<string>):Snap.Element;
+	text(x:number,y:number,text:string|number):Snap.Element;
+	text(x:number,y:number,text:Array<string|number>):Snap.Element;
 	}
 
 	export interface Set {
-		animate(attrs:Object,duration:number,easing?:Function,callback?:Function):Snap.Element;
-        attr(params: Object): Snap.Element;
+		animate(attrs:Object,duration:number,easing?:(num:number)=>number,callback?:()=>void):Snap.Element;
```
Comment (DanielRosenwasser) [inline snapsvg/snapsvg.d.ts]: You should have a different overload; you can omit easing, but not `overload`

File: snapsvg/snapsvg.d.ts [hunk 9]
```diff
@@ -51,13 +73,14 @@ declare module Snap {
...
-	export function animate(from:number,to:Array<number>,setter:Function,duration:number,easing:Function,callback:Function):mina.Mina;
-	export function animate(from:Array<number>,to:Array<number>,setter:Function,duration:number,easing:Function,callback:Function):mina.Mina;
-	export function animation(attr:Object,duration:number,easing?:Function,callback?:Function):Object;
+	export function animate(from:number,to:number,setter:Function,duration:number,easing?:(num:number)=>number,callback?:()=>void):mina.MinaAnimation;
+	export function animate(from:Array<number>,to:number,setter:Function,duration:number,easing?:(num:number)=>number,callback?:()=>void):mina.MinaAnimation;
+	export function animate(from:number,to:Array<number>,setter:Function,duration:number,easing?:(num:number)=>number,callback?:()=>void):mina.MinaAnimation;
+	export function animate(from:Array<number>,to:Array<number>,setter:Function,duration:number,easing?:(num:number)=>number,callback?:()=>void):mina.MinaAnimation;
+	export function animation(attr:Object,duration:number,easing?:(num:number)=>number,callback?:()=>void):Snap.Animation;
```
Comment (DanielRosenwasser) [inline snapsvg/snapsvg.d.ts]: I'm thinking any `attr` object should just be `{ [attrName: string]: number | string | boolean }`

**Comment by lhk** [comment]
Tests 3 : http://svg.dabbles.info/snaptut-blur

There is a sample where the attr object is given a filter. It's "Filters Blur".

There is a sample, where a the method "text" is called on the object returned by group. See: Snap load and animate svg.

The snap documentation is sometimes very confusing. So far, I've left the return types as any.

**Comment by vvakame** [comment]
@lhk thanks mate!

**Pull request merged**